metricbeat/module/linux/perf/perf.go

```diff
@@ -0,0 +1,277 @@
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The linux perf metricset is beta.")
+	linuxModule, ok := base.Module().(*linux.Module)
+	if !ok {
+		return nil, errors.New("unexpected module type")
+	}
+	config := Config{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	if config.SamplePeriod > linuxModule.Period {
+		return nil, fmt.Errorf("Sample period of %s exceeds metricset period of %s", config.SamplePeriod, linuxModule.Period)
+	}
+
+	if config.SamplePeriod == 0 {
+		return nil, fmt.Errorf("Sample Period is zero")
+	}
+
+	var isContinuoius = false
+	if config.SamplePeriod == linuxModule.Period {
+		isContinuoius = true
+	}
+
+	procList, err := matchProcesses(config.Processes)
+	if err != nil {
+		return nil, errors.Wrap(err, "error gathering processes")
+	}
+
+	if len(procList) == 0 {
+		return nil, fmt.Errorf("no processes found that match provided config")
+	}
+
+	if isContinuoius {
+		err := startMonitor(procList)
+		if err != nil {
+			return nil, errors.Wrap(err, "error starting monitor")
+		}
+	}
+
+	// This perf library is Not So Good and eats errors that come from perf_event_open
+	// Do a quck test using a lower-level API to make sure we can actually access perf APIs
+	testP, err := perf.NewMinorFaultsProfiler(procList[0].PID, -1)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating profiler for Minor Page faults")
+	}
+	testP.Close()
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		processes:     procList,
+		period:        config.SamplePeriod,
+		continuous:    isContinuoius,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	perfData, err := runSampleForPeriod(m.processes, m.period, m.continuous)
+	if err != nil {
+		return errors.Wrap(err, "error running sample")
+	}
+
+	for _, process := range perfData {
+
+		event := common.MapStr{}
+
+		if process.SwMetrics.TimeRunning != nil {
+			event["software_events"] = process.SwMetrics
+		}
+
+		if process.HwMetrics.TimeRunning != nil {
+			event["hardware_events"] = process.HwMetrics
+		}
+
+		report.Event(mb.Event{
+			MetricSetFields: event,
+			RootFields: common.MapStr{
+				"process": common.MapStr{
+					"pid":  process.Metadata["pid"],
+					"ppid": process.Metadata["ppid"],
+					"pgid": process.Metadata["pgid"],
+					"name": process.Metadata["name"],
+				},
+			},
+		})
+	}
+
+	return nil
+}
+
+// in cases where we're not stopping monitoring, start before we call `runSampleForPeriod`.
+// Otherwise call every sample period.
+func startMonitor(toSample []procInfo) error {
+	for _, pid := range toSample {
+
+		if pid.HardwareProc != nil {
+			err := pid.HardwareProc.Start()
```

Comment:
In the case where we're not sampling continuously, `startMonitor` will be called as many times as `Fetch()` is called. That means `pid.HardwareProc.Start()` will be called that many times as well. Finally, when the metricset is closed, `pid.HardwareProc.Stop()` will be called once. Same goes for `pid.SoftwareProc`.

Is it safe to call the `Start()` methods several times but only call the `Stop()` methods once?

Reply:
As discussed off-PR, you are already addressing this in `runSampleForPeriod`. Ignore this comment, sorry!